# [CSC8101] Big Data Analytics - 2022 Spark Coursework

## Coursework overview

### Inputs

- **NYC Taxi Trips dataset** - list of recorded taxi trips, each with several characteristics, namely: distance, number of passengers, origin zone, destination zone and trip cost (total amount charged to customer).
- **NYC Zones dataset** - list of zones wherein trips can originate/terminate.

### Tasks

1. Data cleaning
  1. Remove "0 distance" and 'no passengers' records.
  2. Remove outlier records. 
2. Add new columns
  1. Join with zones dataset
  2. Compute the unit profitability of each trip
3. Zone summarisation and ranking
  1. Summarise trip data per zone
  2. Obtain the top 10 ranks according to:
    1. The total trip volume
    2. Their average profitabilitiy
    3. The total passenger volume
4. Record the total and task-specific execution times for each dataset size and format.

### How to

###### Code structure and implementation

- You must implement your solution to each task in the provided function code skeleton.
- The task-specific functions are combined together to form the full pipeline code, executed last (do not modify this code).
- Before implementing the specified function skeleton, you should develop and test your solution on separate code cells (create and destroy cells as needed).

###### Development

- Develop an initial working solution for the 'S' dataset and only then optimise it for larger dataset sizes.
- To perform vectorised operations on a DataFrame:
  - use the API docs to look for existing vectorised functions in: https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql.html#functions
  - if a customised function is required (e.g. to add a new column based on a linear combination of other columns), implement your own User Defined Function (UDF). See:  https://spark.apache.org/docs/latest/sql-ref-functions-udf-scalar.html
- Use only the `pyspark.sql` API - documentation link below - (note that searching through the docs returns results from the `pyspark.sql` API together with the `pyspark.pandas` API):
  - https://spark.apache.org/docs/3.2.0/api/python/reference/pyspark.sql.html
- Periodically download your notebook to your computer as backup and safety measure against accidental file deletion.
 
###### Execution time measurement

- Execution time is calculated and returned by the Spark Engine and shown in the output region of the cell.
- To measure the execution time of a task you must perform a `collect` or similar operation (e.g. `take`) on the returned DataFrame.

## Task 0 - Read data

The code below is ready to run. **Do not modify this code**. It does the following:

- Reads the 'zones' dataset into variable 'zone_names'
- Defines the `init_trips` function that allows you to read the 'trips' dataset (from the DBFS FileStore) given the dataset size ('S' to 'XXL') and format ('parquet' or 'delta') as function arguments
- Defines the `pipeline` function, called in Task 4 to measure the execution time of the entire data processing pipeline
- Shows you how to call the `init_trips` function and display dataset characteristics (number of rows, schema)

In [0]:
## global imports
import pyspark.sql as ps
import pyspark.sql.functions as pf
import pandas as pd

# Load zone names dataset - (much faster to read small file from git than dbfs)
zones_file_url = 'https://raw.githubusercontent.com/NewcastleComputingScience/csc8101-coursework/main/02-assignment-spark/taxi_zone_names.csv'
zone_names = spark.createDataFrame(pd.read_csv(zones_file_url))

# Function to load trips dataset by selected dataset size
def init_trips(size = 'S', data_format = "parquet", taxi_folder = "/FileStore/tables/taxi"):     
    
    files = {
        'S'  : ['2021_07'],
        'M'  : ['2021'],
        'L'  : ['2020_21'],
        'XL' : ['1_6_2019', '7_12_2019'],
        'XXL': ['1_6_2019', '7_12_2019', '2020_21']
    }
    
    # validate input dataset size
    if size not in files.keys():
        print("Invalid input dataset size. Must be one of {}".format(list(files.keys())))
        return None               
    
    if data_format == "parquet":
        filenames = list(map(lambda s: f'{taxi_folder}/parquet/tripdata_{s}.parquet', files[size]))
        trips_df = spark.read.parquet(filenames[0])
        
        for name in filenames[1:]:
            trips_df = trips_df.union(spark.read.parquet(name))
            
    elif data_format == "delta":
        filenames = f"{taxi_folder}/delta/taxi-{size}-delta/"
        trips_df = spark.read.format("delta").load(filenames)
    
    else:
        print("Invalid data format. Must be one of {}".format(['parquet', 'delta']))
        return None
        
    print(
    """
    Trips dataset loaded!
    ---
      Size: {s}
      Format: {f}
      Tables loaded: {ds}
      Number of trips (dataset rows): {tc:,}
    """.format(s = size, f = data_format, ds = filenames, tc = trips_df.count()))
    
    return trips_df

# helper function to print dataset row count
def print_count(df):
    print("Row count: {t:,}".format(t = df.count()))

def pipeline(trips_df, with_task_12 = False, zones_df = zone_names):
    # Do not edit
    #---

    ## Task 1.1
    _trips_11 = t11_remove_zeros(trips_df)

    ## Task 1.2
    if with_task_12:
        _trips_12 = t12_remove_outliers(_trips_11)
    else:
        _trips_12 = _trips_11

    ## Task 2.1
    _trips_21 = t21_join_zones(_trips_12, zones_df = zone_names)

    ## Task 2.2
    _trips_22 = t22_calc_profit(_trips_21)

    ## Task 3.1
    _graph = t31_summarise_trips(_trips_22)

    ## Task 3.2
    _zones = t32_summarise_zones_pairs(_graph)

    _top10_trips     = t32_top10_trips(_zones)
    _top10_profit    = t32_top10_profit(_zones)
    _top10_passenger = t32_top10_passenger(_zones)
    
    return([_top10_trips, _top10_profit, _top10_passenger])

In [0]:
# CHANGE the value of argument 'size' to record the pipeline execution times for increasing dataset sizes
SIZE = 'S'
DATA_FORMAT = 'parquet'


# Load trips dataset
trips = init_trips(SIZE, DATA_FORMAT)

# uncomment line only for small datasets
# trips.take(1)


    Trips dataset loaded!
    ---
      Size: S
      Format: parquet
      Tables loaded: ['/FileStore/tables/taxi/parquet/tripdata_2021_07.parquet']
      Number of trips (dataset rows): 2,898,033
    


In [0]:
print_count(trips)

Row count: 2,898,033


In [0]:
# dataset schemas
trips.printSchema()

root
 |-- index: long (nullable = true)
 |-- VendorID: double (nullable = true)
 |-- tpep_pickup_datetime: string (nullable = true)
 |-- tpep_dropoff_datetime: string (nullable = true)
 |-- passenger_count: double (nullable = true)
 |-- trip_distance: double (nullable = true)
 |-- RatecodeID: double (nullable = true)
 |-- store_and_fwd_flag: string (nullable = true)
 |-- PULocationID: long (nullable = true)
 |-- DOLocationID: long (nullable = true)
 |-- payment_type: double (nullable = true)
 |-- fare_amount: double (nullable = true)
 |-- extra: double (nullable = true)
 |-- mta_tax: double (nullable = true)
 |-- tip_amount: double (nullable = true)
 |-- tolls_amount: double (nullable = true)
 |-- improvement_surcharge: double (nullable = true)
 |-- total_amount: double (nullable = true)
 |-- congestion_surcharge: double (nullable = true)
 |-- cab_type: string (nullable = true)
 |-- lpep_pickup_datetime: string (nullable = true)
 |-- lpep_dropoff_datetime: string (nullable = true)
 |--

In [0]:
display(trips[['PULocationID', 'DOLocationID', 'trip_distance', 'passenger_count', 'total_amount']].take(5))

PULocationID DOLocationID trip_distance passenger_count total_amount 90 68 0.8 1.0 8.8 113 90 0.9 1.0 8.8 88 232 2.8 1.0 13.8 79 249 1.4 1.0 12.3 142 238 2.0 0.0 12.3

In [0]:
zone_names.printSchema()

root
 |-- LocationID: long (nullable = true)
 |-- Borough: string (nullable = true)
 |-- Zone: string (nullable = true)
 |-- service_zone: string (nullable = true)



In [0]:
display(zone_names.take(5))

LocationID Borough Zone service_zone 1 EWR Newark Airport EWR 2 Queens Jamaica Bay Boro Zone 3 Bronx Allerton/Pelham Gardens Boro Zone 4 Manhattan Alphabet City Yellow Zone 5 Staten Island Arden Heights Boro Zone

## Task 1 - Filter rows

**Input:** trips dataset

### Task 1.1 - Remove "0 distance" and 'no passengers' records

Remove dataset rows that represent invalid trips:

- Trips where `trip_distance == 0` (no distance travelled)
- Trips where `passenger_count == 0` and `total_amount == 0` (we want to retain records where `total_amount` > 0 - these may be significant as the taxi may have carried some parcel, for example)

Altogether, a record is removed if it satisfies the following conditions:

`trip_distance == 0` or `(passenger_count == 0` and `total_amount == 0)`.

**Recommended:** Select only the relevant dataset columns for this and subsequent tasks: `['PULocationID', 'DOLocationID', 'trip_distance', 'passenger_count', 'total_amount')]`

### Task 1.2 - Remove outliers using the modified z-score

Despite having removed spurious "zero passengers" trips in task 1.1, columns `total_amount` and `trip_distance` contain additional outlier values that must be identified and removed.

To identify and remove outliers, you will use the modified [z-score](https://en.wikipedia.org/wiki/Standard_score) method.
The modified z-score uses the median and [Median Absolute Deviation](https://en.wikipedia.org/wiki/Median_absolute_deviation) (MAD), instead of the mean and standard deviation, to determine how far an observation (indexed by i) is from the mean:

$$z_i = \frac{x_i - \mathit{median}(\mathbf{x})}{\mathbf{MAD}},$$

where x represents the input vector, xi is an element of x and zi is its corresponding z-score. In turn, the MAD formula is:

$$\mathbf{MAD} = 1.483 * \mathit{median}(\big\lvert x_i - \mathit{median}(\mathbf{x})\big\rvert).$$

Observations with **high** (absolute) z-score are considered outlier observations. A score is considered **high** if its __absolute z-score__ is larger than a threshold T = 3.5:

$$\big\lvert z_i \big\rvert > 3.5.$$

where T represents the number of unit standard deviations beyond which a score is considered an outlier ([wiki](https://en.wikipedia.org/wiki/68%E2%80%9395%E2%80%9399.7_rule)).

This process is repeated twice, once for each of the columns `total_amount` and `trip_distance` (in any order).

**Important:** Use the surrogate function [`percentile_approx`](https://spark.apache.org/docs/3.2.0/api/python/reference/api/pyspark.sql.functions.percentile_approx.html?highlight=percentile#pyspark.sql.functions.percentile_approx) to estimate the median (calculating the median values for a column is expensive as it cannot be parallelised efficiently).

### Task 1.1

In [0]:
# develop your solution here (create/destroy cells as needed) and then implement it in the functions below

In [0]:
# Original Dataset
# display(trips)
display(trips.select('PULocationID', 'DOLocationID', 'trip_distance', 'passenger_count', 'total_amount'))
print_count(trips)

PULocationID DOLocationID trip_distance passenger_count total_amount 90 68 0.8 1.0 8.8 113 90 0.9 1.0 8.8 88 232 2.8 1.0 13.8 79 249 1.4 1.0 12.3 142 238 2.0 0.0 12.3 114 90 1.6 1.0 12.8 90 144 1.8 1.0 13.3 114 48 2.0 1.0 14.75 48 152 5.7 1.0 22.3 234 148 1.8 1.0 14.75 79 141 3.5 2.0 17.3 138 69 7.5 1.0 37.3 230 164 1.02 1.0 10.3 249 79 1.15 1.0 14.16 48 90 2.16 1.0 14.3 90 48 1.07 1.0 9.55 142 238 1.56 1.0 12.36 132 98 8.5 3.0 27.05 132 222 8.4 3.0 27.55 132 74 17.07 1.0 70.6 48 161 0.8 1.0 11.0 107 179 5.1 1.0 25.55 13 246 2.34 1.0 15.96 68 232 5.79 1.0 34.12 138 239 10.0 1.0 49.3 142 239 0.6 1.0 9.95 107 87 4.0 1.0 18.3 231 142 5.0 1.0 24.35 132 158 18.6 1.0 73.1 158 141 3.7 1.0 20.75 75 50 3.56 1.0 17.3 61 74 14.23 5.0 50.8 48 246 1.9 2.0 14.75 50 151 2.9 1.0 16.55 132 90 19.19 2.0 75.72 132 164 16.8 2.0 63.1 137 140 2.1 1.0 11.3 236 74 2.2 1.0 12.74 138 135 6.0 1.0 26.45 132 164 16.8 1.0 78.85 234 255 4.8 1.0 21.8 162 114 2.02 1.0 14.16 114 231 0.97 1.0 10.3 231 144 0.9 1.0 9.3 144 140 5.02 1.0 23.8 238 151 1.27 2.0 10.3 231 262 6.85 1.0 28.8 229 262 1.98 4.0 14.76 162 142 1.57 4.0 12.8 132 188 16.0 1.0 57.65 237 226 2.65 1.0 13.8 100 113 1.8 1.0 12.8 163 143 0.9 1.0 11.15 264 92 4.03 2.0 19.56 138 92 3.8 1.0 16.05 132 42 23.57 5.0 63.1 140 263 0.58 3.0 9.36 263 249 4.93 1.0 19.3 125 141 4.46 2.0 21.36 264 264 1.72 2.0 12.3 264 264 4.91 2.0 25.74 132 230 18.37 2.0 56.55 263 262 0.56 5.0 7.8 132 72 12.11 1.0 46.01 79 90 1.78 2.0 16.56 246 43 6.13 2.0 31.0 239 42 3.22 2.0 15.8 137 146 3.6 2.0 17.8 138 90 11.13 2.0 46.1 79 226 5.68 1.0 22.8 43 151 1.0 1.0 8.16 161 244 8.1 1.0 36.0 164 230 0.9 0.0 10.56 48 239 1.8 0.0 13.8 163 43 18.17 1.0 75.79 138 142 10.64 1.0 51.12 170 141 1.74 1.0 14.12 132 228 22.0 2.0 77.55 132 113 17.99 1.0 75.72 246 74 5.4 1.0 26.16 263 141 0.8 4.0 11.0 141 74 2.1 1.0 11.8 162 263 2.67 1.0 15.36 141 233 1.99 1.0 14.12 114 261 1.64 1.0 12.96 211 246 2.36 1.0 16.56 246 79 2.27 1.0 16.56 230 179 4.63 1.0 24.36 75 237 2.74 1.0 13.8 137 107 0.69 1.0 10.0 79 88 3.9 1.0 19.3 132 61 9.83 1.0 32.05 246 107 1.9 2.0 15.35 170 161 1.2 2.0 10.3 13 48 4.2 1.0 21.95 249 68 1.4 1.0 12.85 48 4 3.11 2.0 20.76 79 140 2.84 3.0 16.62 41 116 2.0 1.0 10.8 116 151 2.0 1.0 14.3 100 249 1.43 1.0 11.76 148 239 5.6 1.0 25.56 100 107 1.53 1.0 -10.8 100 107 1.53 1.0 10.8 79 13 2.83 1.0 15.38 231 186 2.51 2.0 15.8 164 142 2.56 2.0 19.3 161 92 12.6 1.0 49.7 249 87 3.4 1.0 18.3 148 236 6.1 4.0 29.8 234 229 3.07 1.0 18.96 239 239 0.57 1.0 9.96 249 231 1.09 1.0 11.16 79 167 11.21 1.0 -40.3 79 167 11.21 1.0 40.3 229 263 1.96 1.0 12.42 263 142 1.71 1.0 12.96 79 186 2.18 1.0 15.36 249 236 4.6 1.0 22.8 132 100 16.41 5.0 66.85 138 181 13.06 1.0 38.55 231 148 1.36 1.0 14.12 138 113 10.13 1.0 49.19 48 142 0.56 1.0 7.3 230 164 1.5 1.0 11.8 164 75 4.11 1.0 17.8 141 90 3.49 1.0 16.3 137 145 3.5 1.0 18.64 162 42 4.99 1.0 20.8 162 141 0.88 1.0 10.56 163 48 1.59 2.0 14.04 164 158 2.44 1.0 14.5 158 164 1.65 1.0 14.16 164 164 0.63 2.0 10.8 48 74 5.5 2.0 24.36 162 107 1.9 2.0 11.3 79 263 3.73 2.0 15.3 132 95 8.48 1.0 27.55 70 162 8.53 4.0 41.22 132 181 20.31 1.0 58.55 79 249 2.2 1.0 15.8 90 41 5.25 1.0 24.0 163 90 1.68 2.0 10.8 90 13 2.52 2.0 13.3 79 249 1.4 1.0 11.3 132 14 22.91 1.0 62.55 140 212 7.34 2.0 25.8 132 223 14.98 1.0 47.55 148 107 1.53 1.0 12.36 90 50 2.0 1.0 14.8 50 48 0.5 1.0 7.8 90 125 1.5 1.0 10.8 249 229 3.2 1.0 17.05 107 170 0.9 1.0 10.55 79 231 1.6 1.0 11.3 231 230 6.8 1.0 27.3 132 42 18.47 2.0 72.72 161 234 1.27 1.0 12.36 229 137 1.51 2.0 12.36 234 79 1.21 1.0 12.36 79 107 0.78 1.0 9.54 107 170 1.11 1.0 11.8 237 107 2.0 1.0 15.3 68 141 2.8 1.0 18.35 186 25 5.34 2.0 26.62 144 61 5.06 2.0 22.3 234 79 1.0 3.0 9.8 79 114 0.4 2.0 9.35 114 75 5.3 2.0 25.55 75 142 2.5 1.0 16.55 163 239 1.83 1.0 16.62 239 48 1.37 1.0 9.8 48 75 4.17 1.0 20.76 164 11 14.06 1.0 51.36 48 246 1.18 4.0 9.3 246 79 3.08 1.0 20.16 79 189 5.27 1.0 26.76 239 74 3.59 1.0 18.96 68 79 2.3 2.0 17.76 211 255 3.57 1.0 17.8 48 61 10.7 1.0 40.8 61 257 3.1 1.0 13.3 138 76 15

Row count: 2,898,033


In [0]:
# Your solution implementation to task 1.1 goes HERE
def t11_remove_zeros(df):
    # input: trips dataset
    # Select only the relevant columns
    df = df.select('PULocationID', 'DOLocationID', 'trip_distance', 'passenger_count', 'total_amount')
    # Remove trip_distance == 0
    df = df.filter(df.trip_distance != 0)
    # Remove passenger_count == 0 and total_amount == 0, but retain total_amount > 0 -> parcel
    df = df.filter(
        ((df.passenger_count != 0) & (df.total_amount > 0)) | 
        ((df.passenger_count == 0) & (df.total_amount > 0))
    )
    return df

In [0]:
# execute task 1.1
trips_11 = t11_remove_zeros(trips)

print_count(trips_11)

## uncomment only for smaller datasets
# display(trips_11.take(10))

Row count: 2,682,250


In [0]:
# After remove 0 and NO PASSENGER records
display(trips_11)
# display(trips_11.select('PULocationID', 'DOLocationID', 'trip_distance', 'passenger_count', 'total_amount'))
print_count(trips_11)

PULocationID DOLocationID trip_distance passenger_count total_amount 90 68 0.8 1.0 8.8 113 90 0.9 1.0 8.8 88 232 2.8 1.0 13.8 79 249 1.4 1.0 12.3 142 238 2.0 0.0 12.3 114 90 1.6 1.0 12.8 90 144 1.8 1.0 13.3 114 48 2.0 1.0 14.75 48 152 5.7 1.0 22.3 234 148 1.8 1.0 14.75 79 141 3.5 2.0 17.3 138 69 7.5 1.0 37.3 230 164 1.02 1.0 10.3 249 79 1.15 1.0 14.16 48 90 2.16 1.0 14.3 90 48 1.07 1.0 9.55 142 238 1.56 1.0 12.36 132 98 8.5 3.0 27.05 132 222 8.4 3.0 27.55 132 74 17.07 1.0 70.6 48 161 0.8 1.0 11.0 107 179 5.1 1.0 25.55 13 246 2.34 1.0 15.96 68 232 5.79 1.0 34.12 138 239 10.0 1.0 49.3 142 239 0.6 1.0 9.95 107 87 4.0 1.0 18.3 231 142 5.0 1.0 24.35 132 158 18.6 1.0 73.1 158 141 3.7 1.0 20.75 75 50 3.56 1.0 17.3 61 74 14.23 5.0 50.8 48 246 1.9 2.0 14.75 50 151 2.9 1.0 16.55 132 90 19.19 2.0 75.72 132 164 16.8 2.0 63.1 137 140 2.1 1.0 11.3 236 74 2.2 1.0 12.74 138 135 6.0 1.0 26.45 132 164 16.8 1.0 78.85 234 255 4.8 1.0 21.8 162 114 2.02 1.0 14.16 114 231 0.97 1.0 10.3 231 144 0.9 1.0 9.3 144 140 5.02 1.0 23.8 238 151 1.27 2.0 10.3 231 262 6.85 1.0 28.8 229 262 1.98 4.0 14.76 162 142 1.57 4.0 12.8 132 188 16.0 1.0 57.65 237 226 2.65 1.0 13.8 100 113 1.8 1.0 12.8 163 143 0.9 1.0 11.15 264 92 4.03 2.0 19.56 138 92 3.8 1.0 16.05 132 42 23.57 5.0 63.1 140 263 0.58 3.0 9.36 263 249 4.93 1.0 19.3 125 141 4.46 2.0 21.36 264 264 1.72 2.0 12.3 264 264 4.91 2.0 25.74 132 230 18.37 2.0 56.55 263 262 0.56 5.0 7.8 132 72 12.11 1.0 46.01 79 90 1.78 2.0 16.56 246 43 6.13 2.0 31.0 239 42 3.22 2.0 15.8 137 146 3.6 2.0 17.8 138 90 11.13 2.0 46.1 79 226 5.68 1.0 22.8 43 151 1.0 1.0 8.16 161 244 8.1 1.0 36.0 164 230 0.9 0.0 10.56 48 239 1.8 0.0 13.8 163 43 18.17 1.0 75.79 138 142 10.64 1.0 51.12 170 141 1.74 1.0 14.12 132 228 22.0 2.0 77.55 132 113 17.99 1.0 75.72 246 74 5.4 1.0 26.16 263 141 0.8 4.0 11.0 141 74 2.1 1.0 11.8 162 263 2.67 1.0 15.36 141 233 1.99 1.0 14.12 114 261 1.64 1.0 12.96 211 246 2.36 1.0 16.56 246 79 2.27 1.0 16.56 230 179 4.63 1.0 24.36 75 237 2.74 1.0 13.8 137 107 0.69 1.0 10.0 79 88 3.9 1.0 19.3 132 61 9.83 1.0 32.05 246 107 1.9 2.0 15.35 170 161 1.2 2.0 10.3 13 48 4.2 1.0 21.95 249 68 1.4 1.0 12.85 48 4 3.11 2.0 20.76 79 140 2.84 3.0 16.62 41 116 2.0 1.0 10.8 116 151 2.0 1.0 14.3 100 249 1.43 1.0 11.76 148 239 5.6 1.0 25.56 100 107 1.53 1.0 10.8 79 13 2.83 1.0 15.38 231 186 2.51 2.0 15.8 164 142 2.56 2.0 19.3 161 92 12.6 1.0 49.7 249 87 3.4 1.0 18.3 148 236 6.1 4.0 29.8 234 229 3.07 1.0 18.96 239 239 0.57 1.0 9.96 249 231 1.09 1.0 11.16 79 167 11.21 1.0 40.3 229 263 1.96 1.0 12.42 263 142 1.71 1.0 12.96 79 186 2.18 1.0 15.36 249 236 4.6 1.0 22.8 132 100 16.41 5.0 66.85 138 181 13.06 1.0 38.55 231 148 1.36 1.0 14.12 138 113 10.13 1.0 49.19 48 142 0.56 1.0 7.3 230 164 1.5 1.0 11.8 164 75 4.11 1.0 17.8 141 90 3.49 1.0 16.3 137 145 3.5 1.0 18.64 162 42 4.99 1.0 20.8 162 141 0.88 1.0 10.56 163 48 1.59 2.0 14.04 164 158 2.44 1.0 14.5 158 164 1.65 1.0 14.16 164 164 0.63 2.0 10.8 48 74 5.5 2.0 24.36 162 107 1.9 2.0 11.3 79 263 3.73 2.0 15.3 132 95 8.48 1.0 27.55 70 162 8.53 4.0 41.22 132 181 20.31 1.0 58.55 79 249 2.2 1.0 15.8 90 41 5.25 1.0 24.0 163 90 1.68 2.0 10.8 90 13 2.52 2.0 13.3 79 249 1.4 1.0 11.3 132 14 22.91 1.0 62.55 140 212 7.34 2.0 25.8 132 223 14.98 1.0 47.55 148 107 1.53 1.0 12.36 90 50 2.0 1.0 14.8 50 48 0.5 1.0 7.8 90 125 1.5 1.0 10.8 249 229 3.2 1.0 17.05 107 170 0.9 1.0 10.55 79 231 1.6 1.0 11.3 231 230 6.8 1.0 27.3 132 42 18.47 2.0 72.72 161 234 1.27 1.0 12.36 229 137 1.51 2.0 12.36 234 79 1.21 1.0 12.36 79 107 0.78 1.0 9.54 107 170 1.11 1.0 11.8 237 107 2.0 1.0 15.3 68 141 2.8 1.0 18.35 186 25 5.34 2.0 26.62 144 61 5.06 2.0 22.3 234 79 1.0 3.0 9.8 79 114 0.4 2.0 9.35 114 75 5.3 2.0 25.55 75 142 2.5 1.0 16.55 163 239 1.83 1.0 16.62 239 48 1.37 1.0 9.8 48 75 4.17 1.0 20.76 164 11 14.06 1.0 51.36 48 246 1.18 4.0 9.3 246 79 3.08 1.0 20.16 79 189 5.27 1.0 26.76 239 74 3.59 1.0 18.96 68 79 2.3 2.0 17.76 211 255 3.57 1.0 17.8 48 61 10.7 1.0 40.8 61 257 3.1 1.0 13.3 138 76 15.04 1.0 42.3 239 75 2.3 1.0 16.0 249 263 5.1 2

Row count: 2,682,250


### Task 1.2

In [0]:
# Your solution implementation to task 1.2 goes HERE
def t12_remove_outliers(df):
    # Estimate Median of total_amount and trip_distance
    df_medians = df.select(
        pf.percentile_approx("total_amount", 0.5).alias('Median_total_amount'),
        pf.percentile_approx("trip_distance", 0.5).alias('Median_trip_distance')
    )
    df = df.join(df_medians)
    
    # Estimate MAD of total_amount and trip_distance
    ##Calculate 'absolute value of (x - Median)'
    df = df.withColumn(
        "Abs_Diff_total_amount", pf.abs(df.total_amount - df.Median_total_amount)
    ).withColumn(
        "Abs_Diff_trip_distance", pf.abs(df.trip_distance - df.Median_trip_distance)
    )
    ##Calculate 'Median(absolute(x - Median))'
    df_median_abs = df.select(
        pf.percentile_approx("Abs_Diff_total_amount", 0.5).alias('Median_Abs_total_amount'),
        pf.percentile_approx("Abs_Diff_trip_distance", 0.5).alias('Median_Abs_trip_distance')
    )
    df = df.join(df_median_abs)
    ##Calculate MAD
    df = df.withColumn(
        "MAD_total_amount", 1.483 * df.Median_Abs_total_amount
    ).withColumn(
        "MAD_trip_distance", 1.483 * df.Median_Abs_trip_distance
    )
    
    # Estimate absolute z-score of total_amount and trip_distance
    df = df.withColumn(
        "zScore_total_amount", pf.abs(
            (df.total_amount - df.Median_total_amount) / df.MAD_total_amount
        )
    ).withColumn(
        "zScore_trip_distance", pf.abs(
            (df.trip_distance - df.Median_trip_distance)/ df.MAD_trip_distance
        )
    )
    
    # Remove data with absolute z-score larger than 3.5
    df = df.filter(
        (df.zScore_total_amount <= 3.5) & 
        (df.zScore_trip_distance <= 3.5)
    )
    df = df.select('PULocationID', 'DOLocationID', 'trip_distance', 'passenger_count', 'total_amount')
    
    return df

In [0]:
# execute task 1.2
trips_12 = t12_remove_outliers(trips_11)

print_count(trips_12)
# display(trips_12.take(10))

Row count: 2,371,848


In [0]:
display(trips_12)

PULocationID DOLocationID trip_distance passenger_count total_amount 90 68 0.8 1.0 8.8 113 90 0.9 1.0 8.8 88 232 2.8 1.0 13.8 79 249 1.4 1.0 12.3 142 238 2.0 0.0 12.3 114 90 1.6 1.0 12.8 90 144 1.8 1.0 13.3 114 48 2.0 1.0 14.75 48 152 5.7 1.0 22.3 234 148 1.8 1.0 14.75 79 141 3.5 2.0 17.3 230 164 1.02 1.0 10.3 249 79 1.15 1.0 14.16 48 90 2.16 1.0 14.3 90 48 1.07 1.0 9.55 142 238 1.56 1.0 12.36 48 161 0.8 1.0 11.0 107 179 5.1 1.0 25.55 13 246 2.34 1.0 15.96 68 232 5.79 1.0 34.12 142 239 0.6 1.0 9.95 107 87 4.0 1.0 18.3 231 142 5.0 1.0 24.35 158 141 3.7 1.0 20.75 75 50 3.56 1.0 17.3 48 246 1.9 2.0 14.75 50 151 2.9 1.0 16.55 137 140 2.1 1.0 11.3 236 74 2.2 1.0 12.74 138 135 6.0 1.0 26.45 234 255 4.8 1.0 21.8 162 114 2.02 1.0 14.16 114 231 0.97 1.0 10.3 231 144 0.9 1.0 9.3 144 140 5.02 1.0 23.8 238 151 1.27 2.0 10.3 229 262 1.98 4.0 14.76 162 142 1.57 4.0 12.8 237 226 2.65 1.0 13.8 100 113 1.8 1.0 12.8 163 143 0.9 1.0 11.15 264 92 4.03 2.0 19.56 138 92 3.8 1.0 16.05 140 263 0.58 3.0 9.36 263 249 4.93 1.0 19.3 125 141 4.46 2.0 21.36 264 264 1.72 2.0 12.3 264 264 4.91 2.0 25.74 263 262 0.56 5.0 7.8 79 90 1.78 2.0 16.56 246 43 6.13 2.0 31.0 239 42 3.22 2.0 15.8 137 146 3.6 2.0 17.8 79 226 5.68 1.0 22.8 43 151 1.0 1.0 8.16 164 230 0.9 0.0 10.56 48 239 1.8 0.0 13.8 170 141 1.74 1.0 14.12 246 74 5.4 1.0 26.16 263 141 0.8 4.0 11.0 141 74 2.1 1.0 11.8 162 263 2.67 1.0 15.36 141 233 1.99 1.0 14.12 114 261 1.64 1.0 12.96 211 246 2.36 1.0 16.56 246 79 2.27 1.0 16.56 230 179 4.63 1.0 24.36 75 237 2.74 1.0 13.8 137 107 0.69 1.0 10.0 79 88 3.9 1.0 19.3 246 107 1.9 2.0 15.35 170 161 1.2 2.0 10.3 13 48 4.2 1.0 21.95 249 68 1.4 1.0 12.85 48 4 3.11 2.0 20.76 79 140 2.84 3.0 16.62 41 116 2.0 1.0 10.8 116 151 2.0 1.0 14.3 100 249 1.43 1.0 11.76 148 239 5.6 1.0 25.56 100 107 1.53 1.0 10.8 79 13 2.83 1.0 15.38 231 186 2.51 2.0 15.8 164 142 2.56 2.0 19.3 249 87 3.4 1.0 18.3 148 236 6.1 4.0 29.8 234 229 3.07 1.0 18.96 239 239 0.57 1.0 9.96 249 231 1.09 1.0 11.16 229 263 1.96 1.0 12.42 263 142 1.71 1.0 12.96 79 186 2.18 1.0 15.36 249 236 4.6 1.0 22.8 231 148 1.36 1.0 14.12 48 142 0.56 1.0 7.3 230 164 1.5 1.0 11.8 164 75 4.11 1.0 17.8 141 90 3.49 1.0 16.3 137 145 3.5 1.0 18.64 162 42 4.99 1.0 20.8 162 141 0.88 1.0 10.56 163 48 1.59 2.0 14.04 164 158 2.44 1.0 14.5 158 164 1.65 1.0 14.16 164 164 0.63 2.0 10.8 48 74 5.5 2.0 24.36 162 107 1.9 2.0 11.3 79 263 3.73 2.0 15.3 79 249 2.2 1.0 15.8 90 41 5.25 1.0 24.0 163 90 1.68 2.0 10.8 90 13 2.52 2.0 13.3 79 249 1.4 1.0 11.3 148 107 1.53 1.0 12.36 90 50 2.0 1.0 14.8 50 48 0.5 1.0 7.8 90 125 1.5 1.0 10.8 249 229 3.2 1.0 17.05 107 170 0.9 1.0 10.55 79 231 1.6 1.0 11.3 161 234 1.27 1.0 12.36 229 137 1.51 2.0 12.36 234 79 1.21 1.0 12.36 79 107 0.78 1.0 9.54 107 170 1.11 1.0 11.8 237 107 2.0 1.0 15.3 68 141 2.8 1.0 18.35 186 25 5.34 2.0 26.62 144 61 5.06 2.0 22.3 234 79 1.0 3.0 9.8 79 114 0.4 2.0 9.35 114 75 5.3 2.0 25.55 75 142 2.5 1.0 16.55 163 239 1.83 1.0 16.62 239 48 1.37 1.0 9.8 48 75 4.17 1.0 20.76 48 246 1.18 4.0 9.3 246 79 3.08 1.0 20.16 79 189 5.27 1.0 26.76 239 74 3.59 1.0 18.96 68 79 2.3 2.0 17.76 211 255 3.57 1.0 17.8 61 257 3.1 1.0 13.3 239 75 2.3 1.0 16.0 249 263 5.1 2.0 26.16 263 230 2.9 1.0 15.3 233 137 1.51 1.0 14.16 144 164 1.67 1.0 14.75 234 234 0.26 1.0 8.5 113 234 1.3 1.0 12.74 87 261 0.59 1.0 9.36 239 236 1.41 1.0 12.8 144 261 1.41 2.0 10.3 161 236 2.03 1.0 14.76 263 162 2.13 1.0 13.56 164 143 3.03 1.0 17.76 249 161 2.57 1.0 14.3 90 229 2.33 1.0 15.36 164 107 1.05 1.0 10.5 79 4 0.68 1.0 10.56 79 231 1.68 1.0 14.16 211 233 2.87 1.0 17.16 50 163 0.91 2.0 10.56 163 186 1.18 1.0 9.8 43 43 0.75 2.0 8.3 211 90 1.39 1.0 9.8 230 186 1.1 3.0 10.3 234 48 1.4 1.0 10.3 48 90 1.4 1.0 11.3 170 163 1.52 2.0 13.56 162 234 1.4 1.0 11.75 234 233 1.4 1.0 12.35 132 132 0.06 1.0 5.55 239 166 2.6 1.0 16.56 166 74 1.64 1.0 8.8 74 75 1.45 1.0 7.8 87 50 6.26 1.0 24.8 164 262 3.47 2.0 19.56 148 181 6.07 1.0 29.16 24 68 5.08 1.0 21.8 148 90 2.42 2.0 17.76 48 43 1.66 2.0 12.96 161 236 2.31 2.0 12.8 230 170 1.63 1.0 13.56 10

## Task 2 - Compute new columns

### Task 2.1 - Zone names

Obtain the **start** and **end** zone names of each trip by joining the `trips` and `zone_names` datasets (i.e. by using the `zone_names` dataset as lookup table).

**Note:** The columns containing the start and end zone ids of each trip are named `PULocationID` and `DOLocationID`, respectively.

### Task 2.2 - Unit profitability

Compute the column `unit_profitabilty = total_amount / trip_distance`.

### Task 2.1

In [0]:
# develop your solution here (create/destroy cells as needed) and then implement it in the functions below

In [0]:
display(zone_names)

LocationID Borough Zone service_zone 1 EWR Newark Airport EWR 2 Queens Jamaica Bay Boro Zone 3 Bronx Allerton/Pelham Gardens Boro Zone 4 Manhattan Alphabet City Yellow Zone 5 Staten Island Arden Heights Boro Zone 6 Staten Island Arrochar/Fort Wadsworth Boro Zone 7 Queens Astoria Boro Zone 8 Queens Astoria Park Boro Zone 9 Queens Auburndale Boro Zone 10 Queens Baisley Park Boro Zone 11 Brooklyn Bath Beach Boro Zone 12 Manhattan Battery Park Yellow Zone 13 Manhattan Battery Park City Yellow Zone 14 Brooklyn Bay Ridge Boro Zone 15 Queens Bay Terrace/Fort Totten Boro Zone 16 Queens Bayside Boro Zone 17 Brooklyn Bedford Boro Zone 18 Bronx Bedford Park Boro Zone 19 Queens Bellerose Boro Zone 20 Bronx Belmont Boro Zone 21 Brooklyn Bensonhurst East Boro Zone 22 Brooklyn Bensonhurst West Boro Zone 23 Staten Island Bloomfield/Emerson Hill Boro Zone 24 Manhattan Bloomingdale Yellow Zone 25 Brooklyn Boerum Hill Boro Zone 26 Brooklyn Borough Park Boro Zone 27 Queens Breezy Point/Fort Tilden/Riis Beach Boro Zone 28 Queens Briarwood/Jamaica Hills Boro Zone 29 Brooklyn Brighton Beach Boro Zone 30 Queens Broad Channel Boro Zone 31 Bronx Bronx Park Boro Zone 32 Bronx Bronxdale Boro Zone 33 Brooklyn Brooklyn Heights Boro Zone 34 Brooklyn Brooklyn Navy Yard Boro Zone 35 Brooklyn Brownsville Boro Zone 36 Brooklyn Bushwick North Boro Zone 37 Brooklyn Bushwick South Boro Zone 38 Queens Cambria Heights Boro Zone 39 Brooklyn Canarsie Boro Zone 40 Brooklyn Carroll Gardens Boro Zone 41 Manhattan Central Harlem Boro Zone 42 Manhattan Central Harlem North Boro Zone 43 Manhattan Central Park Yellow Zone 44 Staten Island Charleston/Tottenville Boro Zone 45 Manhattan Chinatown Yellow Zone 46 Bronx City Island Boro Zone 47 Bronx Claremont/Bathgate Boro Zone 48 Manhattan Clinton East Yellow Zone 49 Brooklyn Clinton Hill Boro Zone 50 Manhattan Clinton West Yellow Zone 51 Bronx Co-Op City Boro Zone 52 Brooklyn Cobble Hill Boro Zone 53 Queens College Point Boro Zone 54 Brooklyn Columbia Street Boro Zone 55 Brooklyn Coney Island Boro Zone 56 Queens Corona Boro Zone 57 Queens Corona Boro Zone 58 Bronx Country Club Boro Zone 59 Bronx Crotona Park Boro Zone 60 Bronx Crotona Park East Boro Zone 61 Brooklyn Crown Heights North Boro Zone 62 Brooklyn Crown Heights South Boro Zone 63 Brooklyn Cypress Hills Boro Zone 64 Queens Douglaston Boro Zone 65 Brooklyn Downtown Brooklyn/MetroTech Boro Zone 66 Brooklyn DUMBO/Vinegar Hill Boro Zone 67 Brooklyn Dyker Heights Boro Zone 68 Manhattan East Chelsea Yellow Zone 69 Bronx East Concourse/Concourse Village Boro Zone 70 Queens East Elmhurst Boro Zone 71 Brooklyn East Flatbush/Farragut Boro Zone 72 Brooklyn East Flatbush/Remsen Village Boro Zone 73 Queens East Flushing Boro Zone 74 Manhattan East Harlem North Boro Zone 75 Manhattan East Harlem South Boro Zone 76 Brooklyn East New York Boro Zone 77 Brooklyn East New York/Pennsylvania Avenue Boro Zone 78 Bronx East Tremont Boro Zone 79 Manhattan East Village Yellow Zone 80 Brooklyn East Williamsburg Boro Zone 81 Bronx Eastchester Boro Zone 82 Queens Elmhurst Boro Zone 83 Queens Elmhurst/Maspeth Boro Zone 84 Staten Island Eltingville/Annadale/Prince's Bay Boro Zone 85 Brooklyn Erasmus Boro Zone 86 Queens Far Rockaway Boro Zone 87 Manhattan Financial District North Yellow Zone 88 Manhattan Financial District South Yellow Zone 89 Brooklyn Flatbush/Ditmas Park Boro Zone 90 Manhattan Flatiron Yellow Zone 91 Brooklyn Flatlands Boro Zone 92 Queens Flushing Boro Zone 93 Queens Flushing Meadows-Corona Park Boro Zone 94 Bronx Fordham South Boro Zone 95 Queens Forest Hills Boro Zone 96 Queens Forest Park/Highland Park Boro Zone 97 Brooklyn Fort Greene Boro Zone 98 Queens Fresh Meadows Boro Zone 99 Staten Island Freshkills Park Boro Zone 100 Manhattan Garment District Yellow Zone 101 Queens Glen Oaks Boro Zone 102 Queens Glendale Boro Zone 103 Manhattan Governor's Island/Ellis Island/Liberty Island Yellow Zone 104 Manhattan Governor's Island/Ellis Island/Liberty Island Yellow Zone 105 Manhattan

In [0]:
# Your solution implementation to task 2.1 goes HERE
def t21_join_zones(df, zones_df = zone_names):
    # input: output of task 1.2 and zone_names dataset
    # Add Start Zone Names
    df = df.join(zones_df.select('LocationID','Zone'), df.PULocationID == zones_df.LocationID)\
           .withColumnRenamed('Zone','start_zone').drop('LocationID')

    # Add End Zone Names
    df = df.join(zones_df.select('LocationID','Zone'), df.DOLocationID == zones_df.LocationID)\
           .withColumnRenamed('Zone','end_zone').drop('LocationID')

    return df

In [0]:
# execute task 2.1
trips_21 = t21_join_zones(trips_12, zones_df = zone_names)

print_count(trips_21)
# display(trips_21.take(10))

Row count: 2,371,848


In [0]:
display(trips_21)

PULocationID DOLocationID trip_distance passenger_count total_amount start_zone end_zone 90 68 0.8 1.0 8.8 Flatiron East Chelsea 113 90 0.9 1.0 8.8 Greenwich Village North Flatiron 88 232 2.8 1.0 13.8 Financial District South Two Bridges/Seward Park 79 249 1.4 1.0 12.3 East Village West Village 142 238 2.0 0.0 12.3 Lincoln Square East Upper West Side North 114 90 1.6 1.0 12.8 Greenwich Village South Flatiron 90 144 1.8 1.0 13.3 Flatiron Little Italy/NoLiTa 114 48 2.0 1.0 14.75 Greenwich Village South Clinton East 48 152 5.7 1.0 22.3 Clinton East Manhattanville 234 148 1.8 1.0 14.75 Union Sq Lower East Side 79 141 3.5 2.0 17.3 East Village Lenox Hill West 230 164 1.02 1.0 10.3 Times Sq/Theatre District Midtown South 249 79 1.15 1.0 14.16 West Village East Village 48 90 2.16 1.0 14.3 Clinton East Flatiron 90 48 1.07 1.0 9.55 Flatiron Clinton East 142 238 1.56 1.0 12.36 Lincoln Square East Upper West Side North 48 161 0.8 1.0 11.0 Clinton East Midtown Center 107 179 5.1 1.0 25.55 Gramercy Old Astoria 13 246 2.34 1.0 15.96 Battery Park City West Chelsea/Hudson Yards 68 232 5.79 1.0 34.12 East Chelsea Two Bridges/Seward Park 142 239 0.6 1.0 9.95 Lincoln Square East Upper West Side South 107 87 4.0 1.0 18.3 Gramercy Financial District North 231 142 5.0 1.0 24.35 TriBeCa/Civic Center Lincoln Square East 158 141 3.7 1.0 20.75 Meatpacking/West Village West Lenox Hill West 75 50 3.56 1.0 17.3 East Harlem South Clinton West 48 246 1.9 2.0 14.75 Clinton East West Chelsea/Hudson Yards 50 151 2.9 1.0 16.55 Clinton West Manhattan Valley 137 140 2.1 1.0 11.3 Kips Bay Lenox Hill East 236 74 2.2 1.0 12.74 Upper East Side North East Harlem North 138 135 6.0 1.0 26.45 LaGuardia Airport Kew Gardens Hills 234 255 4.8 1.0 21.8 Union Sq Williamsburg (North Side) 162 114 2.02 1.0 14.16 Midtown East Greenwich Village South 114 231 0.97 1.0 10.3 Greenwich Village South TriBeCa/Civic Center 231 144 0.9 1.0 9.3 TriBeCa/Civic Center Little Italy/NoLiTa 144 140 5.02 1.0 23.8 Little Italy/NoLiTa Lenox Hill East 238 151 1.27 2.0 10.3 Upper West Side North Manhattan Valley 229 262 1.98 4.0 14.76 Sutton Place/Turtle Bay North Yorkville East 162 142 1.57 4.0 12.8 Midtown East Lincoln Square East 237 226 2.65 1.0 13.8 Upper East Side South Sunnyside 100 113 1.8 1.0 12.8 Garment District Greenwich Village North 163 143 0.9 1.0 11.15 Midtown North Lincoln Square West 264 92 4.03 2.0 19.56 NV Flushing 138 92 3.8 1.0 16.05 LaGuardia Airport Flushing 140 263 0.58 3.0 9.36 Lenox Hill East Yorkville West 263 249 4.93 1.0 19.3 Yorkville West West Village 125 141 4.46 2.0 21.36 Hudson Sq Lenox Hill West 264 264 1.72 2.0 12.3 NV NV 264 264 4.91 2.0 25.74 NV NV 263 262 0.56 5.0 7.8 Yorkville West Yorkville East 79 90 1.78 2.0 16.56 East Village Flatiron 246 43 6.13 2.0 31.0 West Chelsea/Hudson Yards Central Park 239 42 3.22 2.0 15.8 Upper West Side South Central Harlem North 137 146 3.6 2.0 17.8 Kips Bay Long Island City/Queens Plaza 79 226 5.68 1.0 22.8 East Village Sunnyside 43 151 1.0 1.0 8.16 Central Park Manhattan Valley 164 230 0.9 0.0 10.56 Midtown South Times Sq/Theatre District 48 239 1.8 0.0 13.8 Clinton East Upper West Side South 170 141 1.74 1.0 14.12 Murray Hill Lenox Hill West 246 74 5.4 1.0 26.16 West Chelsea/Hudson Yards East Harlem North 263 141 0.8 4.0 11.0 Yorkville West Lenox Hill West 141 74 2.1 1.0 11.8 Lenox Hill West East Harlem North 162 263 2.67 1.0 15.36 Midtown East Yorkville West 141 233 1.99 1.0 14.12 Lenox Hill West UN/Turtle Bay South 114 261 1.64 1.0 12.96 Greenwich Village South World Trade Center 211 246 2.36 1.0 16.56 SoHo West Chelsea/Hudson Yards 246 79 2.27 1.0 16.56 West Chelsea/Hudson Yards East Village 230 179 4.63 1.0 24.36 Times Sq/Theatre District Old Astoria 75 237 2.74 1.0 13.8 East Harlem South Upper East Side South 137 107 0.69 1.0 10.0 Kips Bay Gramercy 79 88 3.9 1.0 19.3 East Village Financial District South 246 107 1.9 2.0 15.35 West Chelsea/Hudson Yards Gramercy 170 161 1.2 2.0 10.3 Murray Hill Midtown Center 13 48 4.2 1.

### Task 2.2

In [0]:
# Your solution implementation to task 2.2 goes HERE
def t22_calc_profit(df):
    # input: output of task 2.1
    # Compute Unit Profitabilty
    df = df.withColumn("unit_profitability", df.total_amount / df.trip_distance)
    
    return df

In [0]:
# execute task 2.2
trips_22 = t22_calc_profit(trips_21)

print_count(trips_22)
# display(trips_22.take(10))

Row count: 2,371,848


In [0]:
display(trips_22)

PULocationID DOLocationID trip_distance passenger_count total_amount start_zone end_zone unit_profitability 90 68 0.8 1.0 8.8 Flatiron East Chelsea 11.0 113 90 0.9 1.0 8.8 Greenwich Village North Flatiron 9.777777777777779 88 232 2.8 1.0 13.8 Financial District South Two Bridges/Seward Park 4.928571428571429 79 249 1.4 1.0 12.3 East Village West Village 8.785714285714286 142 238 2.0 0.0 12.3 Lincoln Square East Upper West Side North 6.15 114 90 1.6 1.0 12.8 Greenwich Village South Flatiron 8.0 90 144 1.8 1.0 13.3 Flatiron Little Italy/NoLiTa 7.388888888888889 114 48 2.0 1.0 14.75 Greenwich Village South Clinton East 7.375 48 152 5.7 1.0 22.3 Clinton East Manhattanville 3.912280701754386 234 148 1.8 1.0 14.75 Union Sq Lower East Side 8.194444444444445 79 141 3.5 2.0 17.3 East Village Lenox Hill West 4.942857142857143 230 164 1.02 1.0 10.3 Times Sq/Theatre District Midtown South 10.098039215686274 249 79 1.15 1.0 14.16 West Village East Village 12.313043478260871 48 90 2.16 1.0 14.3 Clinton East Flatiron 6.62037037037037 90 48 1.07 1.0 9.55 Flatiron Clinton East 8.925233644859814 142 238 1.56 1.0 12.36 Lincoln Square East Upper West Side North 7.9230769230769225 48 161 0.8 1.0 11.0 Clinton East Midtown Center 13.75 107 179 5.1 1.0 25.55 Gramercy Old Astoria 5.009803921568628 13 246 2.34 1.0 15.96 Battery Park City West Chelsea/Hudson Yards 6.820512820512821 68 232 5.79 1.0 34.12 East Chelsea Two Bridges/Seward Park 5.8929188255613125 142 239 0.6 1.0 9.95 Lincoln Square East Upper West Side South 16.583333333333332 107 87 4.0 1.0 18.3 Gramercy Financial District North 4.575 231 142 5.0 1.0 24.35 TriBeCa/Civic Center Lincoln Square East 4.87 158 141 3.7 1.0 20.75 Meatpacking/West Village West Lenox Hill West 5.608108108108108 75 50 3.56 1.0 17.3 East Harlem South Clinton West 4.859550561797753 48 246 1.9 2.0 14.75 Clinton East West Chelsea/Hudson Yards 7.7631578947368425 50 151 2.9 1.0 16.55 Clinton West Manhattan Valley 5.706896551724139 137 140 2.1 1.0 11.3 Kips Bay Lenox Hill East 5.380952380952381 236 74 2.2 1.0 12.74 Upper East Side North East Harlem North 5.79090909090909 138 135 6.0 1.0 26.45 LaGuardia Airport Kew Gardens Hills 4.408333333333333 234 255 4.8 1.0 21.8 Union Sq Williamsburg (North Side) 4.541666666666667 162 114 2.02 1.0 14.16 Midtown East Greenwich Village South 7.00990099009901 114 231 0.97 1.0 10.3 Greenwich Village South TriBeCa/Civic Center 10.61855670103093 231 144 0.9 1.0 9.3 TriBeCa/Civic Center Little Italy/NoLiTa 10.333333333333334 144 140 5.02 1.0 23.8 Little Italy/NoLiTa Lenox Hill East 4.741035856573705 238 151 1.27 2.0 10.3 Upper West Side North Manhattan Valley 8.11023622047244 229 262 1.98 4.0 14.76 Sutton Place/Turtle Bay North Yorkville East 7.454545454545454 162 142 1.57 4.0 12.8 Midtown East Lincoln Square East 8.152866242038217 237 226 2.65 1.0 13.8 Upper East Side South Sunnyside 5.207547169811321 100 113 1.8 1.0 12.8 Garment District Greenwich Village North 7.111111111111112 163 143 0.9 1.0 11.15 Midtown North Lincoln Square West 12.38888888888889 264 92 4.03 2.0 19.56 NV Flushing 4.853598014888337 138 92 3.8 1.0 16.05 LaGuardia Airport Flushing 4.223684210526316 140 263 0.58 3.0 9.36 Lenox Hill East Yorkville West 16.137931034482758 263 249 4.93 1.0 19.3 Yorkville West West Village 3.9148073022312375 125 141 4.46 2.0 21.36 Hudson Sq Lenox Hill West 4.789237668161435 264 264 1.72 2.0 12.3 NV NV 7.151162790697675 264 264 4.91 2.0 25.74 NV NV 5.242362525458248 263 262 0.56 5.0 7.8 Yorkville West Yorkville East 13.928571428571427 79 90 1.78 2.0 16.56 East Village Flatiron 9.303370786516853 246 43 6.13 2.0 31.0 West Chelsea/Hudson Yards Central Park 5.057096247960849 239 42 3.22 2.0 15.8 Upper West Side South Central Harlem North 4.906832298136646 137 146 3.6 2.0 17.8 Kips Bay Long Island City/Queens Plaza 4.944444444444445 79 226 5.68 1.0 22.8 East Village Sunnyside 4.014084507042254 43 151 1.0 1.0 8.16 Central Park Manhattan Valley 8.16 164 230 0.9 0.0 10.56 Midtown South Times Sq/Theatre 

## Task 3: Rank zones by traffic, passenger volume and profitability

### 3.1 - Summarise interzonal travel

Build a graph data structure of zone-to-zone traffic, representing aggregated data about trips between any two zones. The graph will have one node for each zone and one edge connecting each pair of zones. In addition, edges contain aggregate information about all trips between those zones. 

For example, zones Z1 and Z2 are connected by *two* edges: edge Z1 --> Z2 carries aggregate data about all trips that originated in Z1 and ended in Z2, and edge Z2 --> Z2 carries aggregate data about all trips that originated in Z2 and ended in Z1.

The aggregate information of interzonal travel must include the following data:

- `average_unit_profit` - the average unit profitability (calculated as `mean(unit_profitabilty)`).
- `trips_count` -- the total number of recorded trips.
- `total_passengers` -- the total number of passenger across all trips (sum of `passenger_count`).

This graph can be represented as a new dataframe, with schema:

\[`PULocationID`, `DOLocationID`, `average_unit_profit`, `trips_count`, `total_passengers` \]

__hint__: the `groupby()` operator produces a `pyspark.sql.GroupedData` structure. You can then calculate multiple aggregations from this using `pyspark.sql.GroupedData.agg()`: 
- https://spark.apache.org/docs/3.2.0/api/python/reference/pyspark.pandas/api/pyspark.pandas.DataFrame.groupby.html
- https://spark.apache.org/docs/3.2.0/api/python/reference/api/pyspark.sql.GroupedData.agg.html

### Task 3.2 - Obtain top-10 zones

For each of the following measures, report the top-10 zones _using their plain names you dereferenced in the previous step, not the codes_. Note that this requires ranking the nodes in different orders. Specifically, you need to calculate the following further aggregations:

- the **total** number of trips originating from Z. This is simply the sum of `trips_count` over all outgoing edges for Z, i.e., edges of the form Z -> \*
- the **average** profitability of a zone. This is the average of all `average_unit_profit` over all *outgoing* edges from Z.
- The **total** passenger volume measured as the **sum** of `total_passengers` carried in trips that originate from Z

### Task 3.1

In [0]:
# develop your solution here (create/destroy cells as needed) and then implement it in the functions below

In [0]:
## Your solution to task 3.1 goes HERE
def t31_summarise_trips(df):
    # input: output of task 2.2
    # Compute average_unit_profit, trips_count and total_passengers
    df = df.groupby("PULocationID", "DOLocationID").agg(
        pf.mean("unit_profitability").alias('average_unit_profit'), 
        pf.count("trip_distance").alias('trips_count'), 
        pf.sum("passenger_count").alias('total_passengers')
    )
    
    return df

In [0]:
# execute task 3.1
graph = t31_summarise_trips(trips_22)

print_count(graph)
# display(graph.take(10))

Row count: 10,022


In [0]:
display(graph)

PULocationID DOLocationID average_unit_profit trips_count total_passengers 90 231 7.396050675624754 1094 1576.0 87 33 6.817160729833857 105 145.0 90 142 6.857922958972675 883 1276.0 114 100 7.451603625245185 452 705.0 234 144 9.112465640948626 1465 2186.0 148 262 4.800669891200868 248 398.0 246 249 8.760901617465663 1108 1713.0 170 179 5.109737860724155 76 101.0 48 232 5.427765741716741 317 496.0 142 144 5.754774114280266 326 485.0 231 261 13.014685742613523 487 729.0 193 237 6.9096868993324545 8 10.0 249 225 5.090992513061487 24 31.0 25 255 4.839928177713428 18 24.0 125 255 6.038943405670652 57 86.0 137 37 4.871420114812129 35 48.0 12 111 4.78391959798995 1 1.0 25 148 6.524835050847011 10 16.0 116 127 4.264726008369751 31 54.0 161 193 5.273533412474848 10 12.0 209 190 4.927300835861257 3 3.0 179 260 5.132520744181008 11 27.0 255 262 5.119209039548023 2 2.0 33 36 4.311897746395477 5 6.0 119 263 4.343434343434343 1 5.0 49 234 5.225314733121792 2 2.0 92 138 3.90254892513823 3 5.0 61 34 5.130434782608696 1 1.0 264 36 5.9862700228832955 1 4.0 129 95 4.887713967039971 3 2.0 28 130 7.142272946218802 3 3.0 52 72 4.303030303030303 1 1.0 61 63 4.204470644562811 2 6.0 233 100 11.581744059296007 795 1159.0 143 211 5.497844436957719 101 180.0 235 242 3.339159893420605 3 3.0 25 12 6.0 1 1.0 232 198 4.972081692804949 4 6.0 253 252 3.63905325443787 1 1.0 162 42 4.947142071806624 148 250.0 141 262 11.203822091021422 2804 3913.0 209 40 6.013079259026164 19 28.0 140 141 16.92392945401215 2712 3970.0 132 76 3.6432985799964253 134 184.0 143 151 7.717166542588018 826 1172.0 48 186 10.64832387583625 2482 3776.0 90 164 12.254987020221199 1467 2177.0 41 143 5.9136924808559925 155 212.0 24 24 26.561346519627353 151 208.0 151 163 6.683489748476393 521 738.0 224 50 6.438727788749887 28 38.0 174 47 4.45945945945946 1 4.0 13 66 5.658171687616306 54 86.0 195 181 5.9897294712428915 9 20.0 90 74 4.728009978471466 72 94.0 75 7 4.998190048072055 48 58.0 226 157 4.852790492432403 15 17.0 43 202 4.952082556439153 19 31.0 256 4 7.415875192734478 25 34.0 79 42 4.397867716310721 10 16.0 241 94 5.42713567839196 1 1.0 18 235 6.240965181949164 6 7.0 33 189 6.157840001212693 11 11.0 209 189 6.393449326498894 10 12.0 123 150 5.000000000000001 1 1.0 37 138 3.429951690821256 1 1.0 120 120 6.9850746268656705 1 1.0 114 83 15.645569620253163 1 1.0 147 60 5.615151515151515 2 2.0 74 229 5.423881795447635 88 128.0 233 152 5.27849515625927 5 7.0 141 169 4.241935483870968 1 1.0 209 228 4.814813722531111 5 11.0 230 83 4.920034898840918 9 18.0 45 265 7.12 1 1.0 112 209 4.451510333863275 1 1.0 212 81 3.1627056672760516 1 1.0 79 186 7.886180016480119 2264 3348.0 107 162 8.717809638808461 2564 3633.0 141 43 10.164721569159157 1095 1642.0 125 246 7.1431224464984675 430 635.0 162 246 8.216908056485382 1205 1820.0 161 162 16.950441539432028 2341 3397.0 107 142 6.295911878900256 665 932.0 246 246 19.33903721030715 1660 2625.0 262 74 6.778547382282649 1008 1487.0 209 162 4.656978516398684 90 123.0 264 246 9.191430516586749 75 95.0 233 113 7.535154370421156 413 616.0 237 233 9.24459126166621 1674 2410.0 238 235 3.8594985087538287 17 21.0 142 88 4.79369947304264 54 93.0 116 143 4.905336745635007 65 97.0 24 100 5.586008718112447 58 76.0 125 87 7.917996992797359 207 278.0 234 125 8.621600055074875 606 840.0 249 43 5.961492187322789 338 522.0 107 261 5.9140340838795895 217 342.0 24 50 5.452746473181565 111 165.0 97 87 6.010123756432646 10 13.0 74 162 5.566600322258353 77 100.0 89 89 132.84652325252324 15 19.0 50 262 5.782035627556537 137 201.0 170 66 4.8585601038273944 75 113.0 211 256 6.8207045677169855 90 130.0 48 112 5.566568547204999 76 119.0 97 158 5.6673605640279145 9 13.0 186 152 4.5158785940075346 66 85.0 88 113 6.132046919300501 93 148.0 158 255 6.176554421832048 37 53.0 181 155 4.269293924466338 1 1.0 161 66 4.9903891095827015 16 36.0 12 231 9.199320151933788 43 87.0 225 25 5.482896250192484 8 17.0 148 145 4.941441101200089 63 87.0 4 61 4.7051671732522795 1 1.0 66 13 6.948819513844143

### Task 3.2

In [0]:
# Your solution to task 3.2 goes HERE (implement each of the functions below)
def t32_summarise_zones_pairs(df, zones_df = zone_names):
    # Repeat task 2.1 to recover zone names
    ##Add Start Zone Names
    df = df.join(zones_df.select('LocationID','Zone'), df.PULocationID == zones_df.LocationID)\
           .withColumnRenamed('Zone','start_zone').drop('LocationID')
    ##Add End Zone Names
    df = df.join(zones_df.select('LocationID','Zone'), df.DOLocationID == zones_df.LocationID)\
           .withColumnRenamed('Zone','end_zone').drop('LocationID')
    
    return df

# Top 10 ranked zones by traffic (trip volume)
def t32_top10_trips(df_zones):
    # input: output of task 3.2
    # Compute the Total number of trips
    df_zones = df_zones.groupby("start_zone")\
                       .agg(pf.sum("trips_count").alias('Total_trips'))\
                       .sort("Total_trips", ascending = False)\
                       .limit(10)\
                       .toPandas()
    
    return df_zones

# Top 10 ranked zones by profit
def t32_top10_profit(df_zones):
    # input: output of task 3.2
    # Compute the Average profitability of zone
    df_zones = df_zones.groupby("start_zone")\
                       .agg(pf.avg("average_unit_profit").alias('Average_profitability'))\
                       .sort("Average_profitability", ascending = False)\
                       .limit(10)\
                       .toPandas()
    
    return df_zones

# Top 10 ranked zones by passenger volume
def t32_top10_passenger(df_zones):
    # input: output of task 3.2
    # Compute the Total passenger volume measured as the sum
    df_zones = df_zones.groupby("start_zone")\
                       .agg(pf.sum("total_passengers").alias('Total_passengers'))\
                       .sort("Total_passengers", ascending = False)\
                       .limit(10)\
                       .toPandas()
    
    return df_zones

In [0]:
# execute task 3.2
zones = t32_summarise_zones_pairs(graph)

top10_trips     = t32_top10_trips(zones)
top10_profit    = t32_top10_profit(zones)
top10_passenger = t32_top10_passenger(zones)

In [0]:
display(zones)
print_count(zones)

PULocationID DOLocationID average_unit_profit trips_count total_passengers start_zone end_zone 90 231 7.396050675624754 1094 1576.0 Flatiron TriBeCa/Civic Center 87 33 6.817160729833857 105 145.0 Financial District North Brooklyn Heights 90 142 6.857922958972675 883 1276.0 Flatiron Lincoln Square East 114 100 7.451603625245185 452 705.0 Greenwich Village South Garment District 234 144 9.112465640948626 1465 2186.0 Union Sq Little Italy/NoLiTa 148 262 4.800669891200868 248 398.0 Lower East Side Yorkville East 246 249 8.760901617465663 1108 1713.0 West Chelsea/Hudson Yards West Village 170 179 5.109737860724155 76 101.0 Murray Hill Old Astoria 48 232 5.427765741716741 317 496.0 Clinton East Two Bridges/Seward Park 142 144 5.754774114280266 326 485.0 Lincoln Square East Little Italy/NoLiTa 231 261 13.014685742613523 487 729.0 TriBeCa/Civic Center World Trade Center 193 237 6.9096868993324545 8 10.0 Queensbridge/Ravenswood Upper East Side South 249 225 5.090992513061487 24 31.0 West Village Stuyvesant Heights 25 255 4.839928177713428 18 24.0 Boerum Hill Williamsburg (North Side) 125 255 6.038943405670652 57 86.0 Hudson Sq Williamsburg (North Side) 137 37 4.871420114812129 35 48.0 Kips Bay Bushwick South 12 111 4.78391959798995 1 1.0 Battery Park Green-Wood Cemetery 25 148 6.524835050847011 10 16.0 Boerum Hill Lower East Side 116 127 4.264726008369751 31 54.0 Hamilton Heights Inwood 161 193 5.273533412474848 10 12.0 Midtown Center Queensbridge/Ravenswood 209 190 4.927300835861257 3 3.0 Seaport Prospect Park 179 260 5.132520744181008 11 27.0 Old Astoria Woodside 255 262 5.119209039548023 2 2.0 Williamsburg (North Side) Yorkville East 33 36 4.311897746395477 5 6.0 Brooklyn Heights Bushwick North 119 263 4.343434343434343 1 5.0 Highbridge Yorkville West 49 234 5.225314733121792 2 2.0 Clinton Hill Union Sq 92 138 3.90254892513823 3 5.0 Flushing LaGuardia Airport 61 34 5.130434782608696 1 1.0 Crown Heights North Brooklyn Navy Yard 264 36 5.9862700228832955 1 4.0 NV Bushwick North 129 95 4.887713967039971 3 2.0 Jackson Heights Forest Hills 28 130 7.142272946218802 3 3.0 Briarwood/Jamaica Hills Jamaica 52 72 4.303030303030303 1 1.0 Cobble Hill East Flatbush/Remsen Village 61 63 4.204470644562811 2 6.0 Crown Heights North Cypress Hills 233 100 11.581744059296007 795 1159.0 UN/Turtle Bay South Garment District 143 211 5.497844436957719 101 180.0 Lincoln Square West SoHo 235 242 3.339159893420605 3 3.0 University Heights/Morris Heights Van Nest/Morris Park 25 12 6.0 1 1.0 Boerum Hill Battery Park 232 198 4.972081692804949 4 6.0 Two Bridges/Seward Park Ridgewood 253 252 3.63905325443787 1 1.0 Willets Point Whitestone 162 42 4.947142071806624 148 250.0 Midtown East Central Harlem North 141 262 11.203822091021422 2804 3913.0 Lenox Hill West Yorkville East 209 40 6.013079259026164 19 28.0 Seaport Carroll Gardens 140 141 16.92392945401215 2712 3970.0 Lenox Hill East Lenox Hill West 132 76 3.6432985799964253 134 184.0 JFK Airport East New York 143 151 7.717166542588018 826 1172.0 Lincoln Square West Manhattan Valley 48 186 10.64832387583625 2482 3776.0 Clinton East Penn Station/Madison Sq West 90 164 12.254987020221199 1467 2177.0 Flatiron Midtown South 41 143 5.9136924808559925 155 212.0 Central Harlem Lincoln Square West 24 24 26.561346519627353 151 208.0 Bloomingdale Bloomingdale 151 163 6.683489748476393 521 738.0 Manhattan Valley Midtown North 224 50 6.438727788749887 28 38.0 Stuy Town/Peter Cooper Village Clinton West 174 47 4.45945945945946 1 4.0 Norwood Claremont/Bathgate 13 66 5.658171687616306 54 86.0 Battery Park City DUMBO/Vinegar Hill 195 181 5.9897294712428915 9 20.0 Red Hook Park Slope 90 74 4.728009978471466 72 94.0 Flatiron East Harlem North 75 7 4.998190048072055 48 58.0 East Harlem South Astoria 226 157 4.852790492432403 15 17.0 Sunnyside Maspeth 43 202 4.952082556439153 19 31.0 Central Park Roosevelt Island 256 4 7.415875192734478 25 34.0 Williamsburg (South Side) Alphabet City 79 42 4.397867716310721 10 16.0 East Village Centr

Row count: 10,022


In [0]:
# use 'display()' or return a pandas DataFrame for 'pretty' output
top10_trips

start_zone 
 Total_trips 
 
 
 
 
 0 
 Upper East Side South 
 119874 
 
 
 1 
 Upper East Side North 
 101505 
 
 
 2 
 Penn Station/Madison Sq West 
 95770 
 
 
 3 
 Midtown Center 
 92093 
 
 
 4 
 Murray Hill 
 86251 
 
 
 5 
 Midtown East 
 83735 
 
 
 6 
 Lincoln Square East 
 79180 
 
 
 7 
 Lenox Hill West 
 72533 
 
 
 8 
 Clinton East 
 72509 
 
 
 9 
 Upper West Side South 
 72484

In [0]:
# use 'display()' return a pandas DataFrame for 'pretty' output
top10_profit

start_zone 
 Average_profitability 
 
 
 
 
 0 
 Queensboro Hill 
 259.104384 
 
 
 1 
 Baisley Park 
 193.068838 
 
 
 2 
 Prospect Park 
 169.175210 
 
 
 3 
 Eastchester 
 107.234348 
 
 
 4 
 Saint Michaels Cemetery/Woodside 
 73.315362 
 
 
 5 
 Van Cortlandt Village 
 67.233919 
 
 
 6 
 Crotona Park 
 62.131449 
 
 
 7 
 Riverdale/North Riverdale/Fieldston 
 62.018521 
 
 
 8 
 Flatbush/Ditmas Park 
 58.744308 
 
 
 9 
 South Jamaica 
 48.728374

In [0]:
# use 'display()' or return a pandas DataFrame for 'pretty' output
top10_passenger

start_zone 
 Total_passengers 
 
 
 
 
 0 
 Upper East Side South 
 170039.0 
 
 
 1 
 Upper East Side North 
 144032.0 
 
 
 2 
 Penn Station/Madison Sq West 
 137159.0 
 
 
 3 
 Midtown Center 
 134853.0 
 
 
 4 
 Murray Hill 
 123799.0 
 
 
 5 
 Midtown East 
 119723.0 
 
 
 6 
 Lincoln Square East 
 114752.0 
 
 
 7 
 Clinton East 
 107120.0 
 
 
 8 
 Upper West Side South 
 105888.0 
 
 
 9 
 East Village 
 105571.0

## Task 4 - Record the pipeline's execution time

Record the execution time of:

1. the whole pipeline
2. the whole pipeline except task 1.2

on the two tables below, for all dataset sizes: `'S'`, `'M'`, `'L'`, `'XL'`, `'XXL'`, and data formats: `parquet` and `delta`.

Analyse the resulting execution times and comment on the effect of dataset size, dataset format and task complexity (with and without task 1.2) on pipeline performance.

### Initialise

In [0]:
# after developing your solution, it may be convenient to combine all your functions in a single cell (at the start or end of the notebook)

# CHANGE the value of the following arguments to record the pipeline execution times for increasing dataset sizes
SIZE = 'L'   # S, M, L, XL, XXL
DATA_FORMAT = 'parquet'   # parquet, delta
WITH_TASK_12 = True    # True, False

# Load trips dataset
trips = init_trips(SIZE, DATA_FORMAT)


    Trips dataset loaded!
    ---
      Size: L
      Format: parquet
      Tables loaded: ['/FileStore/tables/taxi/parquet/tripdata_2020_21.parquet']
      Number of trips (dataset rows): 41,953,716
    


### All Solutions

In [0]:
# ALL DEVELOPED SOLUTIONS #

# Task 1.1
def t11_remove_zeros(df):
    # input: trips dataset
    # Select only the relevant columns
    df = df.select('PULocationID', 'DOLocationID', 'trip_distance', 'passenger_count', 'total_amount')
    # Remove trip_distance == 0
    df = df.filter(df.trip_distance != 0)
    # Remove passenger_count == 0 and total_amount <= 0, but retain total_amount > 0 -> parcel
    df = df.filter(
        ((df.passenger_count != 0) & (df.total_amount > 0)) | 
        ((df.passenger_count == 0) & (df.total_amount > 0))
    )
    return df

# Task 1.2
def t12_remove_outliers(df):
    # Estimate Median of total_amount and trip_distance
    df_medians = df.select(
        pf.percentile_approx("total_amount", 0.5).alias('Median_total_amount'),
        pf.percentile_approx("trip_distance", 0.5).alias('Median_trip_distance')
    )
    df = df.join(df_medians)
    
    # Estimate MAD of total_amount and trip_distance
    ##Calculate 'absolute value of (x - Median)'
    df = df.withColumn(
        "Abs_Diff_total_amount", pf.abs(df.total_amount - df.Median_total_amount)
    ).withColumn(
        "Abs_Diff_trip_distance", pf.abs(df.trip_distance - df.Median_trip_distance)
    )
    ##Calculate 'Median(absolute(x - Median))'
    df_median_abs = df.select(
        pf.percentile_approx("Abs_Diff_total_amount", 0.5).alias('Median_Abs_total_amount'),
        pf.percentile_approx("Abs_Diff_trip_distance", 0.5).alias('Median_Abs_trip_distance')
    )
    df = df.join(df_median_abs)
    ##Calculate MAD
    df = df.withColumn(
        "MAD_total_amount", 1.483 * df.Median_Abs_total_amount
    ).withColumn(
        "MAD_trip_distance", 1.483 * df.Median_Abs_trip_distance
    )
    
    # Estimate absolute z-score of total_amount and trip_distance
    df = df.withColumn(
        "zScore_total_amount", pf.abs(
            (df.total_amount - df.Median_total_amount) / df.MAD_total_amount
        )
    ).withColumn(
        "zScore_trip_distance", pf.abs(
            (df.trip_distance - df.Median_trip_distance)/ df.MAD_trip_distance
        )
    )
    
    # Remove data with absolute z-score larger than 3.5
    df = df.filter(
        (df.zScore_total_amount <= 3.5) & 
        (df.zScore_trip_distance <= 3.5)
    )
    df = df.select('PULocationID', 'DOLocationID', 'trip_distance', 'passenger_count', 'total_amount')
    
    return df

# Task 2.1
def t21_join_zones(df, zones_df = zone_names):
    # input: output of task 1.2 and zone_names dataset
    # Add Start Zone Names
    df = df.join(zones_df.select('LocationID','Zone'), df.PULocationID == zones_df.LocationID)\
           .withColumnRenamed('Zone','start_zone').drop('LocationID')

    # Add End Zone Names
    df = df.join(zones_df.select('LocationID','Zone'), df.DOLocationID == zones_df.LocationID)\
           .withColumnRenamed('Zone','end_zone').drop('LocationID')

    return df

# Task 2.2
def t22_calc_profit(df):
    # input: output of task 2.1
    # Compute Unit Profitabilty
    df = df.withColumn("unit_profitability", df.total_amount / df.trip_distance)
    
    return df

# Task 3.1
def t31_summarise_trips(df):
    # input: output of task 2.2
    # Compute average_unit_profit, trips_count and total_passengers
    df = df.groupby("PULocationID", "DOLocationID").agg(
        pf.mean("unit_profitability").alias('average_unit_profit'), 
        pf.count("trip_distance").alias('trips_count'), 
        pf.sum("passenger_count").alias('total_passengers')
    )
    
    return df

# Task 3.2
def t32_summarise_zones_pairs(df, zones_df = zone_names):
    # Repeat task 2.1 to recover zone names
    ##Add Start Zone Names
    df = df.join(zones_df.select('LocationID','Zone'), df.PULocationID == zones_df.LocationID)\
           .withColumnRenamed('Zone','start_zone').drop('LocationID')
    ##Add End Zone Names
    df = df.join(zones_df.select('LocationID','Zone'), df.DOLocationID == zones_df.LocationID)\
           .withColumnRenamed('Zone','end_zone').drop('LocationID')
    
    return df

# Top 10 ranked zones by traffic (trip volume)
def t32_top10_trips(df_zones):
    # input: output of task 3.2
    # Compute the Total number of trips
    df_zones = df_zones.groupby("start_zone")\
                       .agg(pf.sum("trips_count").alias('Total_trips'))\
                       .sort("Total_trips", ascending = False)\
                       .limit(10)\
                       .toPandas()
    
    return df_zones

# Top 10 ranked zones by profit
def t32_top10_profit(df_zones):
    # input: output of task 3.2
    # Compute the Average profitability of zone
    df_zones = df_zones.groupby("start_zone")\
                       .agg(pf.avg("average_unit_profit").alias('Average_profitability'))\
                       .sort("Average_profitability", ascending = False)\
                       .limit(10)\
                       .toPandas()
    
    return df_zones

# Top 10 ranked zones by passenger volume
def t32_top10_passenger(df_zones):
    # input: output of task 3.2
    # Compute the Total passenger volume measured as the sum
    df_zones = df_zones.groupby("start_zone")\
                       .agg(pf.sum("total_passengers").alias('Total_passengers'))\
                       .sort("Total_passengers", ascending = False)\
                       .limit(10)\
                       .toPandas()
    
    return df_zones

### Run & Record

In [0]:
# run and record the resulting execution time shown by databricks (on the cell footer)

# IMPORTANT: this function calls all task functions in order of occurrence. For this code to run without errors, you have to load into memory all of the previous task-specific functions, even if you haven't implemented these yet.
pipeline(trips, with_task_12 = WITH_TASK_12)

Out[81]: [                     start_zone  Total_trips
 0         Upper East Side South      1819890
 1         Upper East Side North      1715429
 2                Midtown Center      1334351
 3  Penn Station/Madison Sq West      1250953
 4                  Midtown East      1179117
 5           Lincoln Square East      1142439
 6                   Murray Hill      1121997
 7         Upper West Side South      1099328
 8               Lenox Hill West      1074113
 9                  Clinton East      1013937,
                          start_zone  Average_profitability
 0                    Newark Airport              91.735409
 1                  Sunset Park East              90.008716
 2                          Canarsie              76.334374
 3                         Glen Oaks              72.828478
 4                Murray Hill-Queens              62.618637
 5                          Rosedale              56.228450
 6                       JFK Airport              53.389259
 7  

_Table 1. Pipeline performance for `parquet` format._

| metric                      | S    | M    | L    | XL   | XXL  |
|-----------------------------|------|------|------|------|------|
| rows (M)                    | 2.9  | 15.6 |  42.0 |  90.4 |  132.4 |
| execution time   (w/o 1.2)  | 7.00sec | 1.61min | 13.54min | 19.16min | 28.92min |
| execution time              | 23.02sec | 4.59min | 36.94min | 50.91min | 35.04min |
| sec / 1M records (w/o 1.2)  | 2.41  | 6.19  | 19.34  | 12.72  | 13.11  |
| sec / 1M records            | 7.94  | 17.65  | 52.77  | 33.79 | 15.88  |

_Table 2. Pipeline performance for `delta` format._

| metric                      | S    | M    | L    | XL   | XXL  |
|-----------------------------|------|------|------|------|------|
| rows (M)                    |  2.9 |  15.6 |  42.0 |  90.4 |  132.4 |
| execution time   (w/o 1.2)  | 7.24sec | 10.45sec | 9.54sec | 10.61sec | 10.33sec |
| execution time              | 12.60sec | 29.44sec | 31.18sec | 49.52sec | 1.07min |
| sec / 1M records (w/o 1.2)  | 2.50  | 0.67  | 0.23  | 0.12  | 0.08  |
| sec / 1M records            | 4.34  | 1.89  | 0.74  | 0.55  | 0.48  |